File: src/js/src/loading/IDataSetLoader.ts
import type { IFlatFileReader } from '../interfaces/IFlatFileReader';
import type { DataSources, ProgressFn } from '../interfaces/IApiSdk';
import type {
  Voyage,
  Ship,
  CabinGrade,
  CabinOffering,
  Departure,
  Port,
} from '../data';

/**
 * The fully-built object graph produced by a {@link IDataSetLoader}. These are
 * exactly the collections and lookup maps that {@link ApiSdk} commits to its
 * private fields, so a loader's output can be assigned wholesale.
 */
export interface DataSetResult {
  voyages: Voyage[];
  ships: Ship[];
  cabinGrades: CabinGrade[];
  ports: Port[];
  departures: Departure[];
  offerings: CabinOffering[];

  shipById: Map<string, Ship>;
  cabinGradeByCode: Map<string, CabinGrade>;
  portByCode: Map<string, Port>;
  departureByCode: Map<string, Departure>;
}

/**
 * Strategy for turning a set of flat files into the navigable object graph.
 * One implementation exists per flat-file {@link DataSourceFormat} ("v1",
 * "v3"). {@link ApiSdk.load} selects the right loader and assigns its result.
 *
 * This abstraction mirrors the parallel .NET implementation 1:1 (same name and
 * semantics) so the two SDKs stay aligned.
 */
export interface IDataSetLoader {
  /**
   * Read the flat files via {@link reader} and assemble the wired graph. The
   * returned collections/maps are linked in both directions so the result is
   * traversable from any entity.
   */
  load(
    reader: IFlatFileReader,
    sources: DataSources,
    onProgress: ProgressFn
  ): Promise<DataSetResult>;
}

File: src/dotnet/ApiSdk/Loading/IDataSetLoader.cs
namespace ApiSdk.Loading;

/// <summary>
/// Builds the navigable object graph from a set of flat files. Each flat-file
/// format (dev, prod) has its own implementation; <see cref="ApiSdk.LoadAsync"/>
/// dispatches to the right one based on <see cref="DataSources.Format"/> and
/// assigns the returned <see cref="DataSetLoadResult"/> onto its fields.
/// </summary>
internal interface IDataSetLoader
{
    /// <summary>
    /// Read the files referenced by <paramref name="sources"/> through
    /// <paramref name="fileReader"/> and assemble the fully-wired graph.
    /// </summary>
    Task<DataSetLoadResult> LoadAsync(
        IFlatFileReader fileReader,
        DataSources sources,
        IProgress<string>? progress);
}
